icicle/backend/cpu/src/ring/cpu_norm.cpp

```diff
@@ -0,0 +1,358 @@
+
+// Get number of threads from config (defined in cpu_vec_ops.cpp)
+int get_nof_workers(const VecOpsConfig& config);
+
+using uint128_t = __uint128_t;
+
+static int64_t abs_centered(int64_t val, int64_t q)
+{
+  if (val > q / 2) { val = q - val; }
+  return val;
+}
+
+static bool validate_input_range(int64_t val, int64_t sqrt_q)
+{
+  if (val >= sqrt_q) {
+    ICICLE_LOG_ERROR << "Input value " << val << " is greater than sqrt(q) = " << sqrt_q;
+    return false;
+  }
+  return true;
+}
+
+static eIcicleError cpu_check_norm_bound(
+  const Device& device,
+  const field_t* input,
+  size_t size,
+  eNormType norm,
+  uint64_t norm_bound,
+  const VecOpsConfig& config,
+  bool* output)
+{
+  if (!input || !output) {
+    ICICLE_LOG_ERROR << "Invalid pointer: null pointer.";
+    return eIcicleError::INVALID_POINTER;
+  }
+
+  if (size > 65536) { // size of the element shouldn't be bigger than 2^16
+    ICICLE_LOG_ERROR << "Invalid argument: vector size must be at most 65536.";
+    return eIcicleError::INVALID_ARGUMENT;
+  }
+
+  if (size == 0) {
+    for (uint32_t i = 0; i < config.batch_size; ++i) {
+      output[i] = true;
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  if (config.columns_batch) {
+    ICICLE_LOG_ERROR << "Unsupported config: norm check does not support column batch.";
+    return eIcicleError::INVALID_ARGUMENT;
+  }
+
+  constexpr auto q_storage = field_t::get_modulus();
+  const int64_t q = *(const int64_t*)&q_storage;
+
+  if (q < 0) {
+    ICICLE_LOG_ERROR << "Field modulus q must be less than 64 bits; received q = " << q;
+    return eIcicleError::INVALID_ARGUMENT;
+  }
+
+  auto sqrt_q = static_cast<uint32_t>(std::sqrt(q));
+
+  const int64_t* input_i64 = reinterpret_cast<const int64_t*>(input);
+
+  // For L2 norm, we need to check the ||input||² < norm_bound²
+  if (norm == eNormType::L2) {
+    const uint128_t bound_squared = static_cast<uint128_t>(norm_bound) * static_cast<uint128_t>(norm_bound);
+    const int nof_workers = get_nof_workers(config);
+    std::vector<std::vector<uint128_t>> thread_sums(config.batch_size, std::vector<uint128_t>(nof_workers, 0));
+    std::atomic<bool> validation_failed(false);
+
+    tf::Taskflow taskflow;
+    tf::Executor executor;
+    const uint64_t worker_task_size = (size + nof_workers - 1) / nof_workers;
+
+    for (uint64_t start_idx = 0, thread_idx = 0; start_idx < size; start_idx += worker_task_size, ++thread_idx) {
+      taskflow.emplace([=, &thread_sums, &validation_failed]() {
+        const uint64_t end_idx = std::min(start_idx + worker_task_size, static_cast<uint64_t>(size));
+        std::vector<uint128_t> local_sums(config.batch_size, 0);
+
+        for (uint64_t idx = start_idx; idx < end_idx; ++idx) {
+          for (uint32_t batch_idx = 0; batch_idx < config.batch_size; ++batch_idx) {
+            int64_t val = input_i64[batch_idx * size + idx];
+            if (!validate_input_range(val, sqrt_q)) {
```

Comment:
Either validating the input should be done after centering OR validating should check that the input will be valid post centering:

```cpp
static bool validate_input_range(int64_t val, int64_t sqrt_q)
{
  if (sqrt_q <= val && val <= (q - sqrt_q)) {
    ICICLE_LOG_ERROR << "Input value " << val << " is greater than sqrt(q) = " << sqrt_q;
    return false;
  }
  return true;
}
```